src/libpmemobj/critnib.c

```diff
@@ -0,0 +1,462 @@
+	if (!n)
+		return UNLOCK, NULL;
+	uint64_t del = util_fetch_and_add64(&c->write_status, 1) % DELETED_LIFE;
+	free_node(c, c->pending_dels[del][0]);
+	free_node(c, c->pending_dels[del][1]);
+	c->pending_dels[del][1] = c->pending_dels[del][0] = 0;
+	if (n->shift == ENDBIT) {
+		if (n->path == key && n != &nullnode) {
+			util_atomic_store_explicit64((uint64_t *)&c->root,
+				(uint64_t)&nullnode, memory_order_release);
+			void *value = n->child[0];
+			c->pending_dels[del][0] = n;
+			return UNLOCK, value;
+		}
+		return UNLOCK, NULL;
+	}
+	struct critnib_node **k_parent = &c->root, **n_parent = &c->root,
+			    *k = n;
+
+	while (k->shift != ENDBIT) {
+		n_parent = k_parent;
+		n = k;
+		k_parent = &k->child[(key >> k->shift) & NIB];
+		k = *k_parent;
+	}
+	if (k->path != key)
+		return UNLOCK, NULL;
+
+	util_atomic_store_explicit64(
+		(uint64_t *)&n->child[(key >> n->shift) & NIB],
+		(uint64_t)&nullnode, memory_order_release);
+
+	int ochild = -1;
+	for (int i = 0; i < SLNODES; i++)
+		if (n->child[i] != &nullnode) {
+			if (ochild != -1) {
+				void *value = k->child[0];
+				c->pending_dels[del][0] = k;
+				return UNLOCK, value;
+			} else
+				ochild = i;
+		}
+	if (ochild == -1)
+		ochild = 0;
+
+	util_atomic_store_explicit64((uint64_t *)n_parent,
+		(uint64_t)n->child[ochild], memory_order_release);
+	void *value = k->child[0];
+	c->pending_dels[del][0] = n;
+	c->pending_dels[del][1] = k;
+	return UNLOCK, value;
+}
+
+/*
+ * critnib_get -- query for a key ("==" match), returns value or NULL
+ *
+ * Doesn't need a lock but if many deletes happened while our thread was
+ * somehow stalled the query is restarted (as freed nodes remain unused only
+ * for a grace period).
+ *
+ * Counterintuitively, it's pointless to return the most current answer,
+ * we need only one that was valid at any point after the call started.
+ */
+void *
+critnib_get(struct critnib *c, uint64_t key)
+{
+	uint64_t wrs1, wrs2;
+	void *res;
+
+	do {
+		util_atomic_load_explicit64(&c->write_status, &wrs1,
+			memory_order_acquire);
+		struct critnib_node *n = c->root;
+		if (!n)
+			return 0;
+		/*
+		 * critbit algorithm: dive into the tree, looking at nothing but
+		 * each node's critical bit^H^H^Hnibble.  This means we risk
+		 * going wrong way if our path is missing, but that's ok...
+		 */
+		while (n->shift != ENDBIT)
+			n = n->child[(key >> n->shift) & NIB];
+		/* ... as we check it at the end. */
+		res = (n->path == key) ? n->child[0] : 0;
+		util_atomic_load_explicit64(&c->write_status, &wrs2,
+			memory_order_acquire);
+	} while (wrs1 + DELETED_LIFE <= wrs2);
+	return res;
+}
+
+/*
+ * internal: find_le_dive -- return the rightmost non-null node in a subtree
+ */
+static void *
+find_le_dive(struct critnib_node *__restrict n, uint64_t key)
+{
+deeper:
+	for (int nib = NIB; nib >= 0; nib--)
+		if (n->child[nib] != &nullnode) {
+			n = n->child[nib];
+			if (n->shift == ENDBIT)
+				return n->child[0];
+			goto deeper;
```

Comment:
I would use recursion (or just a `for` loop, which I think is possible here) here instead. But that's my personal preference, I'll let others give an opinion.